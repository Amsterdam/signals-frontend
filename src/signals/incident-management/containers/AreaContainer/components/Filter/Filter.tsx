import { Heading, Icon, themeSpacing, List, ListItem } from '@amsterdam/asc-ui'
import styled from 'styled-components'

import configuration from 'shared/services/configuration/configuration'
import { dateToString } from 'shared/services/date-utils'

interface FilterProps {
  subcategory?: string
  departments?: string
  startDate: string
}

const ICON_SIZE = 20

const Wrapper = styled.section`
  padding: ${themeSpacing(4)};
`

const Field = styled.div`
  margin-top: ${themeSpacing(4)};
`

const Title = styled.div`
  font-weight: bold;
  margin-bottom: ${themeSpacing(2)};
`

const StyledIcon = styled(Icon)`
  display: inline-block;
  margin-right: ${themeSpacing(4)};
  position: relative;
  top: 4px;
`

const Filter: React.FC<FilterProps> = (props) => {
  const subcategory = props.subcategory ? (
    <Field>
      <Title data-testid="subcategory-label">
        Subcategorie (verantwoordelijke afdeling)
      </Title>
      <List>
        <ListItem>
          <span data-testid="subcategory">{props.subcategory} </span>
          <span data-testid="departments">({props.departments})</span>
        </ListItem>
      </List>
    </Field>
  ) : null

  return (
    <Wrapper>
      <Heading>Filter</Heading>
      {subcategory}
      <Field>
        <Title data-testid="status-label">Status</Title>
        <List>
          <ListItem>
            <StyledIcon size={ICON_SIZE}>
              <img
                alt="Groene locatie pin"
                src="/assets/images/area-map/icon-pin.svg"
              />
            </StyledIcon>
            Openstaand
          </ListItem>
          <ListItem>
            <StyledIcon size={ICON_SIZE}>
              <img
                alt="Zwarte locatie pin"
                src="/assets/images/area-map/icon-pin-green.svg"
              />
            </StyledIcon>
            Afgehandeld
          </ListItem>
        </List>
      </Field>

      <Field>
        <Title data-testid="period-label">Periode</Title>
        <List>
          <ListItem data-testid="period">
            Van {dateToString(new Date(props.startDate))} t/m NU
          </ListItem>
        </List>
      </Field>

      <Field>
        <Title data-testid="area-label">Omgeving</Title>
        <List>
          <ListItem>
            <StyledIcon size={ICON_SIZE}>
              <img
                alt="Draadkruis"
                src="/assets/images/area-map/icon-cross-small.svg"
              />
            </StyledIcon>
            Locatie huidige melding
          </ListItem>
          <ListItem>
            <StyledIcon size={ICON_SIZE}>
              <img alt="Straal" src="/assets/images/area-map/icon-radius.svg" />
            </StyledIcon>
            {`Straal ${configuration.map.optionsAreaMap.focusRadiusMeters}m`}
          </ListItem>
        </List>
      </Field>
      <Field>
        <Title forwardedAs="h4" data-testid="kind-label">
          Soort
        </Title>
        <List>
          <ListItem>Standaardmelding</ListItem>
          <ListItem>Deelmelding</ListItem>
        </List>
      </Field>
    </Wrapper>
  )
}

export default Filter
