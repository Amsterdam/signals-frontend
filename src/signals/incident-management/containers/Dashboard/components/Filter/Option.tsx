// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { Dispatch, SetStateAction } from 'react'
import { useEffect, useRef } from 'react'

import { Controller, useFormContext } from 'react-hook-form'

import { OptionLi } from './styled'
import type { Filter } from './types'
import type { Option as OptionType } from './types'

type Props = {
  activeFilter: Filter
  setFilterNameActive: (name: string) => void
  option: OptionType
  index: number
  focus: number
  setFocus: Dispatch<SetStateAction<number>>
}

export const Option = ({
  activeFilter,
  setFocus,
  setFilterNameActive,
  option,
  index,
  focus,
}: Props) => {
  const { control, getValues } = useFormContext()

  const optionRef = useRef<HTMLLIElement | null>(null)

  useEffect(() => {
    if (focus === index) {
      optionRef.current?.focus()
    }
  }, [focus, index])

  return (
    <Controller
      key={`${option}${index}`}
      name={activeFilter.name}
      control={control}
      render={({ field: { onChange, ref } }) => (
        <OptionLi
          selected={option.value === getValues(activeFilter.name)?.value}
          role="option"
          tabIndex={0}
          ref={(el) => {
            ref(el)
            optionRef.current = el
          }}
          onClick={() => {
            onChange(option)
            setFilterNameActive('')
          }}
          onKeyDown={(e) => {
            e.preventDefault()
            setFocus(index)

            if (['Enter', 'Space'].includes(e.code)) {
              onChange(option)
              setFilterNameActive('')
            } else if (e.code === 'Escape') {
              setFilterNameActive('')
            }
          }}
        >
          {option.display}
        </OptionLi>
      )}
    />
  )
}
