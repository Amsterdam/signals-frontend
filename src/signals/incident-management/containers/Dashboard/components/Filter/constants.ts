// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
export const filterNames = [
  'department',
  'category',
  'priority',
  'punctuality',
  'district',
]
