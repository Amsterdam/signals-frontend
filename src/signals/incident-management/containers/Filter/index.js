import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose, bindActionCreators } from 'redux';

import { makeSelectEditFilter } from 'signals/incident-management/selectors';
import FilterForm from 'signals/incident-management/components/FilterForm';
import * as types from 'shared/types';

import {
  applyFilter,
  clearEditFilter as onClearFilter,
  filterEditCanceled,
  filterSaved as onSaveFilter,
  filterUpdated as onUpdateFilter,
} from 'signals/incident-management/actions';

export const FilterContainerComponent = ({
  onApplyFilter,
  onCancel,
  onFilterEditCancel,
  onSubmit,
  filter,
  ...rest
}) => {
  const editFilter = useMemo(() => filter, [filter]);

  const onFormSubmit = useCallback(
    (event, filterData) => {
      onApplyFilter(filterData);
      onSubmit(event);
    },
    [onApplyFilter, onSubmit]
  );

  const onEditCancel = useCallback(() => {
    onFilterEditCancel();
    onCancel();
  }, [onFilterEditCancel, onCancel]);

  return <FilterForm {...rest} filter={editFilter} onCancel={onEditCancel} onSubmit={onFormSubmit} />;
};

FilterContainerComponent.propTypes = {
  filter: types.filterType.isRequired,
  onApplyFilter: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onClearFilter: PropTypes.func.isRequired,
  onFilterEditCancel: PropTypes.func.isRequired,
  onSaveFilter: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onUpdateFilter: PropTypes.func.isRequired,
};

const mapStateToProps = () =>
  createStructuredSelector({
    filter: makeSelectEditFilter,
  });

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      onApplyFilter: applyFilter,
      onClearFilter,
      onFilterEditCancel: filterEditCanceled,
      onSaveFilter,
      onUpdateFilter,
    },
    dispatch
  );

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(FilterContainerComponent);
