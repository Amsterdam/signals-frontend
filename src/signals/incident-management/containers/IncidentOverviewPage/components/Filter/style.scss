@import '~stijl/dist/scss/ams-colorpalette';

.filter-component {
  padding: 10px;

  &__title {
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 10px;
  }

  &__body {
    border-top: 1px solid black;
    padding-top: 5px;
  }
}
