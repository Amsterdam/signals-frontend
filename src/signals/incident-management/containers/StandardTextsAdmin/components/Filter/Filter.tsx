// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import Label from 'components/Label'
import RadioButtonList from 'components/RadioButtonList'
import type { RadioButtonOption } from 'components/RadioButtonList/types'
import { changeStatusOptionList } from 'signals/incident-management/definitions/statusList'

import { Form } from './styled'
import type {
  Option,
  StandardTextsAdminValue,
  StandardTextsData,
  Facet,
} from '../../types'

const getStatusOptions = (options: RadioButtonOption[], meta?: Facet[]) => {
  return options.map((option) => {
    const optionMeta = meta?.find((data) => data.term === option.key)

    const count = optionMeta?.count ? { count: optionMeta?.count } : {}

    return {
      key: option.key,
      value: option.value,
      ...count,
    }
  })
}

const getActiveOptions = (meta?: Facet[]) => {
  return [
    {
      key: 'true',
      value: 'Actief',
    },
    {
      key: 'false',
      value: 'Non-actief',
    },
  ].map((option) => {
    const optionMeta = meta?.find((data) => data.term.toString() === option.key)
    const count = optionMeta?.count ? { count: optionMeta?.count } : {}

    return {
      ...option,
      ...count,
    }
  })
}
export interface Props {
  setActiveFilter: (active: Option) => void
  setStatusFilter: (filter: Option) => void
  currentStatusFilter: StandardTextsAdminValue['statusFilter']
  currentActiveFilter: StandardTextsAdminValue['activeFilter']
  meta?: StandardTextsData['facets']
}

export const Filter = ({
  setStatusFilter,
  setActiveFilter,
  currentStatusFilter,
  currentActiveFilter,
  meta,
}: Props) => {
  const onStatusChange = (_groupName: string, option: RadioButtonOption) => {
    setStatusFilter(option)
  }

  const onActiveChange = (_groupName: string, option: RadioButtonOption) => {
    setActiveFilter(option)
  }

  const statusOptions = getStatusOptions(changeStatusOptionList, meta?.state)
  const activeOptions = getActiveOptions(meta?.active)

  return (
    <Form>
      <Label as="span">Filter op status</Label>
      <RadioButtonList
        emptySelectionLabel="Alle statussen"
        groupName="filter-status"
        options={statusOptions}
        onChange={onStatusChange}
        defaultValue={currentStatusFilter?.key}
      />

      <Label as="span">Filter op actief/non-actief</Label>
      <RadioButtonList
        emptySelectionLabel="Alle"
        groupName="filter-on-active"
        options={activeOptions}
        onChange={onActiveChange}
        defaultValue={currentActiveFilter?.key}
      />
    </Form>
  )
}
