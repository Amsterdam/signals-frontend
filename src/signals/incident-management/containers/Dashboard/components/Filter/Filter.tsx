// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { useCallback, useEffect, useRef, useState } from 'react'

import { useClickOutside } from '@amsterdam/asc-ui'
import { useForm, FormProvider } from 'react-hook-form'

import useLocationReferrer from 'hooks/useLocationReferrer'
import { generateParams } from 'shared/services/api/api'
import { useIncidentManagement } from 'signals/incident-management/context'
import { INCIDENTS_URL } from 'signals/incident-management/routes'

import { filterNames } from './constants'
import SelectList from './SelectList'
import { FilterBar } from './styled'
import type { DashboardFilter } from './types'

type Props = {
  setQueryString: (queryString: string) => void
}

export const Filter = ({ setQueryString }: Props) => {
  const [filterActiveName, setFilterActiveName] = useState('')
  const { dashboardFilter, setDashboardFilter } = useIncidentManagement()

  const location = useLocationReferrer() as { referrer: string }

  // TODO: check whether INCIDENT URL is needed here.
  const methods = useForm<DashboardFilter>({
    defaultValues: Object.fromEntries(
      filterNames.map((name) => [
        name,
        (location?.referrer === INCIDENTS_URL &&
          dashboardFilter &&
          dashboardFilter[name]) || {
          display: '',
          value: '',
        },
      ])
    ),
  })

  const { getValues, watch, resetField } = methods

  const refFilterContainer = useRef<HTMLDivElement>(null)

  useClickOutside(
    refFilterContainer,
    () => filterActiveName && setFilterActiveName('')
  )

  const updateFilters = useCallback(() => {
    const filters = getValues()

    setDashboardFilter((prevFilter) => ({
      ...prevFilter,
      ...filters,
    }))

    const selectedFilters: { [key: string]: string } = Object.fromEntries(
      Object.entries(filters).map(([key, value]) => {
        return [[key], value?.value]
      })
    )

    const params = generateParams(selectedFilters)

    setQueryString(params)
  }, [getValues, setDashboardFilter, setQueryString])

  useEffect(() => {
    const subscription = watch((_, { name, type }) => {
      if (type === 'change') {
        updateFilters()

        if (name === 'department') {
          resetField('category_slug')
        }
      }
    })

    return () => {
      subscription.unsubscribe()
    }
  }, [getValues, updateFilters, resetField, watch])

  useEffect(() => {
    refFilterContainer.current?.scrollIntoView({
      behavior: 'smooth',
      block: 'end',
    })
  }, [filterActiveName])

  useEffect(() => {
    updateFilters()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <FilterBar ref={refFilterContainer}>
      <FormProvider {...methods}>
        <SelectList
          filterActiveName={filterActiveName}
          setFilterActiveName={setFilterActiveName}
          setQueryString={setQueryString}
        />
      </FormProvider>
    </FilterBar>
  )
}
