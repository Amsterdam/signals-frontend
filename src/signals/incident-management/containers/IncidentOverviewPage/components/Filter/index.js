import React from 'react';
import PropTypes from 'prop-types';
import { FormBuilder, FieldGroup } from 'react-reactive-form';

import './style.scss';
import { TextInput } from '../../../../components/TextInput';
import { SelectInput } from '../../../../components/SelectInput';

class Filter extends React.Component {
  constructor(props) {
    super(props);
    if (this.props.filter) {
      this.filterForm.setValue(props.filter);
    }
  }

  onFilter = (filter) => {
    this.props.filterIncidents({ filter });
  }

  filterForm = FormBuilder.group({
    id: [''],
    incident_date_start: [''],
    location__stadsdeel: [['']],
    category__sub: [['']],
    status__state: [['']],
    location__address_text: [''],
  });

  handleReset = () => {
    this.filterForm.reset();
    this.onFilter(this.filterForm.value);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.onFilter(this.filterForm.value);
  }

  render() {
    const { subcategoryList, statusList, stadsdeelList } = this.props;
    return (
      <div className="filter-component">
        <div className="filter-component__title">Filters</div>
        <div className="filter-component__body">
          <FieldGroup
            control={this.filterForm}
            render={({ invalid }) => (
              <form onSubmit={this.handleSubmit}>
                <div>
                  <TextInput name="id" display="Id" control={this.filterForm.get('id')} />
                  <TextInput name="incident_date_start" display="Datum" control={this.filterForm.get('incident_date_start')} placeholder={'JJJJ-MM-DD'} />
                  <SelectInput name="location__stadsdeel" display="Stadsdeel" control={this.filterForm.get('location__stadsdeel')} values={stadsdeelList} multiple />
                  <SelectInput name="category__sub" display="Rubriek" control={this.filterForm.get('category__sub')} values={subcategoryList} multiple size={10} />
                  <SelectInput name="status__state" display="Status" control={this.filterForm.get('status__state')} values={statusList} multiple />
                  <TextInput name="location__address_text" display="Adres" control={this.filterForm.get('location__address_text')} />

                  <button className="action tertiair" onClick={this.handleReset} type="button">
                    <span className="value">Reset filter</span>
                  </button>
                  <button className="action primary" type="submit" disabled={invalid}>
                    <span className="value">Zoek</span>
                  </button>
                </div>
              </form>
            )}
          />
        </div>
      </div>
    );
  }
}

Filter.propTypes = {
  stadsdeelList: PropTypes.array,
  subcategoryList: PropTypes.array,
  statusList: PropTypes.array,
  filter: PropTypes.object,
  filterIncidents: PropTypes.func.isRequired
};

export default Filter;
