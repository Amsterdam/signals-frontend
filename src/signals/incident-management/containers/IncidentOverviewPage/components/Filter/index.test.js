import React from 'react';
import { shallow } from 'enzyme';

import Filter from './index';

describe('<Filter />', () => {
  let renderedComponent;
  let props;

  beforeEach(() => {
    props = {
      filterIncidents: jest.fn()
    };

    renderedComponent = shallow(
      <Filter {...props} />
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    expect(renderedComponent).toMatchSnapshot();
  });

  // it('should render 2 buttons', () => {
  //   console.log(renderedComponent.debug());
  //   expect(renderedComponent.find('button').length).toEqual(2);
  // });
});
