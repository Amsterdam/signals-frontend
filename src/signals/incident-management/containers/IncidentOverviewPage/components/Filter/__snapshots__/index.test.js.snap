// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> should render correctly 1`] = `
<div
  className="filter-component"
>
  <div
    className="filter-component__title"
  >
    Filters
  </div>
  <div
    className="filter-component__body"
  >
    <FieldGroup
      control={
        FormGroup {
          "_onDisabledChange": Array [],
          "_pendingChange": false,
          "_pendingDirty": false,
          "_pendingTouched": false,
          "asyncValidator": null,
          "controls": Object {
            "category__sub": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": Array [
                "",
              ],
              "asyncValidator": null,
              "errors": null,
              "formState": Array [
                "",
              ],
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": Array [
                "",
              ],
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
            "id": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": "",
              "asyncValidator": null,
              "errors": null,
              "formState": "",
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": "",
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
            "incident_date_start": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": "",
              "asyncValidator": null,
              "errors": null,
              "formState": "",
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": "",
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
            "location__address_text": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": "",
              "asyncValidator": null,
              "errors": null,
              "formState": "",
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": "",
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
            "location__stadsdeel": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": Array [
                "",
              ],
              "asyncValidator": null,
              "errors": null,
              "formState": Array [
                "",
              ],
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": Array [
                "",
              ],
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
            "status__state": FormControl {
              "_onCollectionChange": [Function],
              "_onDisabledChange": Array [],
              "_parent": [Circular],
              "_pendingChange": true,
              "_pendingDirty": false,
              "_pendingTouched": false,
              "_pendingValue": Array [
                "",
              ],
              "asyncValidator": null,
              "errors": null,
              "formState": Array [
                "",
              ],
              "get": [Function],
              "getError": [Function],
              "handler": [Function],
              "hasError": [Function],
              "meta": Object {},
              "onBlur": [Function],
              "onBlurChanges": Observable {
                "observers": Array [],
              },
              "onChange": [Function],
              "onValueChanges": Observable {
                "observers": Array [],
              },
              "patchValue": [Function],
              "pristine": true,
              "reset": [Function],
              "setValue": [Function],
              "stateChanges": Observable {
                "observers": Array [],
              },
              "status": "VALID",
              "statusChanges": Observable {
                "observers": Array [],
              },
              "submitted": false,
              "touched": false,
              "validator": null,
              "validatorsOrOpts": Object {
                "asyncValidators": null,
                "updateOn": null,
                "validators": null,
              },
              "value": Array [
                "",
              ],
              "valueChanges": Observable {
                "observers": Array [],
              },
            },
          },
          "errors": null,
          "get": [Function],
          "getError": [Function],
          "handleSubmit": [Function],
          "hasError": [Function],
          "meta": Object {},
          "patchValue": [Function],
          "pristine": true,
          "reset": [Function],
          "setValue": [Function],
          "stateChanges": Observable {
            "observers": Array [],
          },
          "status": "VALID",
          "statusChanges": Observable {
            "observers": Array [],
          },
          "submitted": false,
          "touched": false,
          "validator": null,
          "validatorOrOpts": Object {
            "asyncValidators": null,
            "updateOn": null,
            "validators": null,
          },
          "value": Object {
            "category__sub": Array [
              "",
            ],
            "id": "",
            "incident_date_start": "",
            "location__address_text": "",
            "location__stadsdeel": Array [
              "",
            ],
            "status__state": Array [
              "",
            ],
          },
          "valueChanges": Observable {
            "observers": Array [],
          },
        }
      }
      render={[Function]}
      strict={true}
    />
  </div>
</div>
`;
